subsys/storage/flash_buffered_write/fbw.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2017, 2020 Nordic Semiconductor ASA
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME FBW
+#define LOG_LEVEL CONFIG_FBW_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, CONFIG_FBW_LOG_LEVEL);
+
+#include <zephyr/types.h>
+#include <string.h>
+#include <drivers/flash.h>
+
+#include <storage/fbw.h>
+
+#ifdef CONFIG_FBW_ERASE
+
+int fbw_erase(struct fbw_ctx *ctx, off_t off)
+{
+	int rc;
+	struct flash_pages_info page;
+
+	rc = flash_get_page_info_by_offs(ctx->fdev, off, &page);
+	if (rc != 0) {
+		LOG_ERR("Error %d while getting page info", rc);
+		return rc;
+	}
+
+	if (ctx->last_erased_page_start_offset == page.start_offset) {
+		return 0;
+	}
+
+	ctx->last_erased_page_start_offset = page.start_offset;
+	LOG_INF("Erasing page at offset 0x%08lx", (long)page.start_offset);
+
+	flash_write_protection_set(ctx->fdev, false);
+	rc = flash_erase(ctx->fdev, page.start_offset, page.size);
+	flash_write_protection_set(ctx->fdev, true);
+
+	if (rc != 0) {
+		LOG_ERR("Error %d while erasing page", rc);
+	}
+
+	return rc;
+}
+
+#endif /* CONFIG_FBW_ERASE */
+
+static int flash_sync(struct fbw_ctx *ctx)
+{
+	int rc = 0;
+	size_t write_addr = ctx->offset + ctx->bytes_written;
+
+
+	if (IS_ENABLED(CONFIG_FBW_ERASE)) {
+		rc = fbw_erase(ctx, write_addr + ctx->buf_bytes);
+		if (rc < 0) {
+			LOG_ERR("fbw_erase error %d offset=0x%08zx", rc,
+					write_addr);
+			return rc;
+		}
+	}
+
+	flash_write_protection_set(ctx->fdev, false);
+	rc = flash_write(ctx->fdev, write_addr, ctx->buf, ctx->buf_bytes);
+	flash_write_protection_set(ctx->fdev, true);
+
+	if (rc != 0) {
+		LOG_ERR("flash_write error %d offset=0x%08zx", rc,
+			write_addr);
+		return rc;
+	}
+
+	if (ctx->callback) {
+		rc = flash_read(ctx->fdev, write_addr, ctx->buf,
+				ctx->buf_bytes);
+		if (rc != 0) {
+			LOG_ERR("flash read failed: %d", rc);
+			return rc;
+		}
+
+		rc = ctx->callback(ctx->buf, ctx->buf_bytes, write_addr);
+		if (rc != 0) {
+			LOG_ERR("callback failed: %d", rc);
+		}
+	}
+
+	ctx->bytes_written += ctx->buf_bytes;
+	ctx->buf_bytes = 0U;
+
+	return rc;
+}
+
+int fbw_write(struct fbw_ctx *ctx, const u8_t *data, size_t len, bool flush)
+{
+	int processed = 0;
+	int rc = 0;
+	int buf_empty_bytes;
+
+	if (!ctx || !data) {
+		return -EFAULT;
+	}
+
+	if (ctx->bytes_written + ctx->buf_bytes + len > ctx->available) {
+		return -ENOMEM;
+	}
+
+	while ((len - processed) >=
+	       (buf_empty_bytes = ctx->buf_len - ctx->buf_bytes)) {
+		memcpy(ctx->buf + ctx->buf_bytes, data + processed,
```

Comment:
What's the rationale for copying the data out of the provided region in order to initiate the write, rather than just doing the write from where the data currently lives?  Is there some promise that a round-robin input buffer can wrap into the being-written data without corrupting the material written?

Reply:
There are a couple of reasons:
- Write alignment. This is not guaranteed by the drivers, and the user of `fbw` should not have to deal with it.
- Avoid synchronization issues with radio. For some HW there are restrictions on flash operations working in parallel with radio operations, the fewer flash operations the less intrusive the buffered write is to the radio performance.
- Fewer write operations will give faster average response time as seen from the app (since flash_write takes longer than memcpy).
- Fewer write operations also implies fewer calls to the driver which gives a faster application using less power
- Backwards compatibility, mcumgr transport was tested with this and there might be timing related issues for systems already using flash_img.c if they suddenly stop buffering their writes.